analyzers/src/SonarAnalyzer.CSharp/Rules/PrivateStaticMethodUsedOnlyByNestedClass.cs

```diff
@@ -102,25 +102,18 @@ private static bool IsPrivateAndStatic(MethodDeclarationSyntax method, TypeDecla
 
     private static TypeDeclarationSyntax LowestCommonAncestorOrSelf(IEnumerable<TypeDeclarationSyntax> declaredTypes)
     {
-        var typeHierarchyFromTopToBottom = declaredTypes.Select(PathFromTop);
+        var typeHierarchyFromTopToBottom = declaredTypes.Select(PathFromTop).ToArray();
         var minPathLength = typeHierarchyFromTopToBottom.Select(x => x.Length).Min();
         var firstPath = typeHierarchyFromTopToBottom.First();
+        var levels = Enumerable.Range(0, minPathLength);
+        var firstDifferingLevel = levels.FirstOrDefault(NotEveryNodeIsTheSameOnLevel);
+        int levelOfLowestCommonAncestor = firstDifferingLevel == default
+            ? minPathLength - 1
+            : firstDifferingLevel - 1;
+        return firstPath[levelOfLowestCommonAncestor];
 
-        var lastCommonPathIndex = 0;
-        for (int i = 0; i < minPathLength; i++)
-        {
-            var isPartOfCommonPath = typeHierarchyFromTopToBottom.All(x => x[i] == firstPath[i]);
-            if (isPartOfCommonPath)
-            {
-                lastCommonPathIndex = i;
-            }
-            else
-            {
-                break;
-            }
-        }
-
-        return firstPath[lastCommonPathIndex];
+        bool NotEveryNodeIsTheSameOnLevel(int level) =>
```

Comment:
I don't find this clearer because it's two negations:
- NotEveryNodeIsTheSameOnLevel
- `!=` inside Any

Also, `levels` refers to indeces, whereas `NotEveryNodeIsTheSameOnLevel` is iterating over `typeHierarchyFromTopToBottom`

But `NotEveryNodeIsTheSameOnLevel` is called on `levels`

There's a lot of indirection in here.

Reply:
Reverted it. Now the only change is splitting the comment to multiple lines.